pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,269 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
-			iterationBegin = time.Now()
-			encodeDuration = time.Duration(0)
-			sendDuration   = time.Duration(0)
-			seriesCount    int
-			chunksCount    int
-		)
-
-		// Once the iteration is done we will update the stats.
-		defer stats.update(func(stats *queryStats) {
-			stats.mergedSeriesCount += seriesCount
-			stats.mergedChunksCount += chunksCount
-
-			// The time spent iterating over the series set is the
-			// actual time spent fetching series and chunks, encoding and sending them to the client.
-			// We split the timings to have a better view over how time is spent.
-			stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration
-			stats.streamingSeriesEncodeResponseDuration += encodeDuration
-			stats.streamingSeriesSendResponseDuration += sendDuration
-			stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
-				stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
-		})
+		err = s.sendSeriesChunks(req, srv, seriesSet, stats)
+		if err != nil {
+			return
+		}
+	})
 
-		for seriesSet.Next() {
-			var lset labels.Labels
-			var series storepb.Series
+	if err != nil {
+		return
+	}
 
-			seriesCount++
+	if req.StreamingChunksBatchSize == 0 || req.SkipChunks {
+		// Hints and stats were not sent before, so send it now.
+		return s.sendHintsAndStats(srv, resHints, stats)
+	}
 
-			// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
-			// because the subsequent call to seriesSet.Next() may release it.
-			if req.SkipChunks {
-				lset, _ = seriesSet.At()
-			} else {
-				lset, series.Chunks = seriesSet.At()
+	return nil
+}
 
-				chunksCount += len(series.Chunks)
-				s.metrics.chunkSizeBytes.Observe(float64(chunksSize(series.Chunks)))
-			}
-			series.Labels = mimirpb.FromLabelsToLabelAdapters(lset)
+// sendStreamingSeriesLabelsHintsStats sends the labels of the streaming series.
+// Since hints and stats need to be sent before the "end of stream" streaming series message,
+// this function also sends the hints and the stats.
+func (s *BucketStore) sendStreamingSeriesLabelsHintsStats(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	stats *safeQueryStats,
+	seriesSet storepb.SeriesSet,
+	resHints *hintspb.SeriesResponseHints,
+) error {
+	// TODO: should we pool the seriesBuffer/seriesBatch?
+	seriesBuffer := make([]*storepb.StreamingSeries, req.StreamingChunksBatchSize)
+	for i := range seriesBuffer {
+		seriesBuffer[i] = &storepb.StreamingSeries{}
+	}
+	seriesBatch := &storepb.StreamSeriesBatch{
+		Series: seriesBuffer[:0],
+	}
+	// TODO: can we send this in parallel while we start fetching the chunks below?
+	for seriesSet.Next() {
+		var lset labels.Labels
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		// TODO: check if it is safe to hold the lset.
```

Comment:
It is safe because we currently don't pool the slices of labels themselves. We pool the slices of series, where each series has one label set

https://github.com/grafana/mimir/blob/df0a7ea985a730149eeb4ce9766659520e65cec9/pkg/storegateway/series_refs.go#L96-L120

You can change the IMPORTANT comment to mention this. And it would be great if this is covered by tests. You can maybe try to modify the lsets from the pool in `newSeriesChunkRefsSet` and see if tests break (they should).

But at the same time pooling slices of labels is probably a big potential win because we can reduce number of allocations drastically.

If you want to be super safe you can change `sendStreamingSeriesLabelsHintsStats` to work with `seriesChunkRefsSetIterator` instead of `storepb.SeriesSet`. This way `sendStreamingSeriesLabelsHintsStats` will have control over when to release or not the loaded `seriesChunkRefsSet`.

Reply:
I think it is covered by tests because I am getting race error below where I am sending chunks and re-using the memory (fixing that now).

Reply:
> But at the same time pooling slices of labels is probably a big potential win because we can reduce number of allocations drastically.

Not sure I understand this. Do you mean pool `[]labels.Labels` (i.e. a slice of series, which I think we are already doing?), or pool `labels.Labels` (i.e. a single series, hence a slice of individual label). If it is the latter, we cannot do it, because of the labels implementation has been redesigned to use a single string.

Reply:
> Not sure I understand this. Do you mean pool `[]labels.Labels` (i.e. a slice of series, which I think we are already doing?), or pool `labels.Labels` (i.e. a single series, hence a slice of individual label). If it is the latter, we cannot do it, because of the labels implementation has been redesigned to use a single string.

you're right, we can't pool `labels.Labels` anymore. We can still potentially pool the underlying byte slices of the new labels-as-string implementation and avoid having to allocate a new string for each series.


I only attempted to pool `labels.Labels` once and it showed fairly good results, but then gave up because labels-as-strings were being introduced shortly after. labels-as-string is still in progress and pooling them might happen only after some time. In this case I guess it's ok to only update the comment in this function